lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -389,4 +501,353 @@ private static ParquetCorruptionException corruptionException(String column, Par
+        return filter;
+    }
+
+    /**
+     * This class implements methods defined in UserDefinedPredicate based on the page statistic and tuple domain(for a column).
+     */
+    static class DomainUserDefinedPredicate<T extends Comparable<T>>
+            extends UserDefinedPredicate<T>
+            implements Serializable
+    {
+        private Domain columnDomain;
+
+        DomainUserDefinedPredicate(Domain domain)
+        {
+            this.columnDomain = domain;
+        }
+
+        @Override
+        public boolean keep(T value)
+        {
+            if (value == null && !columnDomain.isNullAllowed()) {
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public boolean canDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistic)
+        {
+            if (statistic == null) {
+                return false;
+            }
+            else {
+                if (statistic.getMin() instanceof Integer) {
+                    Integer min = (Integer) statistic.getMin();
+                    Integer max = (Integer) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics((long) min, (long) max));
+                }
+                else if (statistic.getMin() instanceof Long) {
+                    Long min = (Long) statistic.getMin();
+                    Long max = (Long) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics(min, max));
+                }
+                else if (statistic.getMin() instanceof Float) {
+                    Integer min = floatToRawIntBits((Float) statistic.getMin());
+                    Integer max = floatToRawIntBits((Float) statistic.getMax());
+                    return canDropCanWithRangeStats(new ParquetIntegerStatistics((long) min, (long) max));
+                }
+                else if (statistic.getMin() instanceof Double) {
+                    Double min = (Double) statistic.getMin();
+                    Double max = (Double) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetDoubleStatistics(min, max));
+                }
+                else if (statistic.getMin() instanceof Binary && columnDomain.getType() instanceof VarcharType) {
+                    Binary min = (Binary) statistic.getMin();
+                    Binary max = (Binary) statistic.getMax();
+                    return canDropCanWithRangeStats(new ParquetStringStatistics((Slices.wrappedBuffer(min.getBytes())), Slices.wrappedBuffer(max.getBytes())));
+                }
+                //TODO: Add other types
+            }
+            return false;
+        }
+
+        @Override
+        public boolean inverseCanDrop(org.apache.parquet.filter2.predicate.Statistics<T> statistics)
+        {
+            // !canDrop() cannot be used because it might not be correct. To be safe, we just keep the record by returning false.
```

Comment:
Why might `!canDrop` be incorrect?

Also, I'd split this comment onto two lines after the period.